LX-24 Adds management command to copy Units from Open edX to Blockstore

pomegranited opened the pull request:
**Description:** Adds a Open edX Django plugin application for transferring data between Open edX and Blockstore.

Initial commit provides a management command which transfers a Unit (vertical) to a Blockstore Bundle.

**JIRA:** LX-24, OC-5133

**Dependencies:**

* https://github.com/edx/edx-platform/pull/18855 -- for static asset parsing
* https://github.com/open-craft/blockstore/pull/4 -- for API
* https://github.com/open-craft/blockstore/pull/5 -- for bundle structure

**Installation instructions:**

1. Install this branch in the CMS shell.
1. Run the Blockstore service so the CMS shell can access its REST API.
1. Ensure that the URL for the Blockstore is present in `settings.BLOCKSTORE_API_URL` (see [`openedx_blockstore_relay.settings`](https://github.com/open-craft/openedx-blockstore-relay/tree/jill/transfer-unit/openedx_blockstore_relay/settings)).

**Testing instructions:**

1. Using the Blockstore REST API, create a Collection, and note its UUID.
   You can pass the `--collection-uuid` to the command below, and a new Bundle will be created for each transfer.
   Alternatively, you can create a Bundle and pass its `--bundle-uuid` instead.
1. Test uploading the various supported types of blocks to the Collection: `course`, `chapter`, `sequential`, `vertical`, `problem`, `video`, `html`.
   Unsupported block types (e.g. `discussion`) are skipped, and show a warning.
   Usage example:
   ```
   ./manage.py cms transfer_to_blockstore --settings=devstack \
      --block-key "block-v1:edX+DemoX+Demo_Course+type@vertical+block@256f17a44983429fb1a60802203ee4e0" \
      --collection-uuid "cccccccc-cccc-cccc-cccc-cccccccccccc"
   ```
1. Visit the Blockstore Bundle API to see the new/updated bundles.
   Each transfer should store in the Bundle:
   * An `.olx` file for the top-level block and each child block
   * All static course assets referenced by the block and its children (with `public=True`).
      HTML blocks also upload a separate `.html` file (with `public=False`).
   * A `bundle.json` with:
      * The `components` list will contain only the top-level `.olx` path.
      * The `type` field will reflect the type of the top-level block.
      * The `assets` and `dependencies` lists will be empty.

**Author Notes & Concerns:**

1. Because we have to upload each block file individually to the bundle, a full snapshot of the Bundle is made at each upload step.
   It would be nice to be able to limit the number of snapshots taken, since these intermediate snapshots will be inherently incomplete and may be invalid.

**Reviewers:**
- [ ] @symbolist

CC @ormsbee @bradenmacdonald 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

pomegranited committed:
Initial requirements

Generated from `make upgrade`

pomegranited committed:
Fixes quality warnings

pomegranited committed:
Update tox targets

pomegranited committed:
Start management command

Validate the arguments and pass them to the class that will handle the data transfer.

pomegranited committed:
Configure as CMS django plugin app

pomegranited committed:
Initial commit of management command and unit transfer class

pomegranited committed:
Generalize the data transfer function

* renamed data transfer management command to allow for sequence blocks
  (eventually, not implemented yet)
* tried to consolidate the HTML content into the OLX block
* trying to post multiple files -- not working
* fixes quality issues

pomegranited commented on openedx_blockstore_relay/data_transfer/unit.py:
```diff
@@ -0,0 +1,116 @@
+
+        TODO consider requests.toolbelt to stream large file content: https://toolbelt.readthedocs.io/en/latest/
+        """
+        # FIXME requests says it is possible to post more than one file, but the bundle isn't ending up with the right
```
The `POST` below fails with a `ValueError: too many values to unpack` when I run this code using a DemoX HTML block, e.g.:
```bash
./manage.py cms transfer_to_blockstore --settings=devstack --verbosity=3 --block "block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4" --bundle "9a3bb1e0-463d-4972-9a06-a433b3dfd872"
```

debug log shows:
```
DEBUG|OLX: <html filename="030e35c4756a4ddc8d40b95fbbfff4d4" />
DEBUG|POST files=[(u'data', (u'030e35c4756a4ddc8d40b95fbbfff4d4.html', <fs.iotools.RawWrapper object at 0x7f7fe1b3c1d0>)), (u'data', (u'030e35c4756a4ddc8d40b95fbbfff4d4.olx', <fs.iotools.RawWrapper object at 0x7f7fe1b3c310>))], data=[{u'path': u'/html/030e35c4756a4ddc8d40b95fbbfff4d4.html', u'data': u'030e35c4756a4ddc8d40b95fbbfff4d4.html', u'public': True}, {u'path': u'/html/030e35c4756a4ddc8d40b95fbbfff4d4.olx', u'data': u'030e35c4756a4ddc8d40b95fbbfff4d4.olx', u'public': True}] to Bundle <http://localhost:8888/api/v1/bundles/9a3bb1e0-463d-4972-9a06-a433b3dfd872/files/>
```

pomegranited committed:
Writes out the bundle.json file

Removes consolidate_olx since that wasn't working.

pomegranited committed:
Cleaned up unit upload code

* ensured that bundle.json has a "type", a "main_component", and optionally a "components" list.
* avoided iterating over the files more than required
* clarified public/private methods

pomegranited committed:
Added support for storing verticals

and consolidated the code further

pomegranited committed:
Renames transfer module to transfer_data

pomegranited committed:
Made an explicit choice not to support course/section/subsection blocks yet

pomegranited committed:
Adds method to add Studio runtime to a given block descriptor

Will be needed when fetching static assets for a block.

pomegranited committed:
As per the bundle.js format and conventions:
* Store only the top-level OLX path in "components"
* Store empty "assets" and "dependencies" lists.

pomegranited committed:
Uses future.moves to import urllib

pomegranited committed:
Adds static assets to bundle

* uses change from https://github.com/edx/edx-platform/pull/18855 to scrape
  static assets from the OLX and related files
* requires a running LMS/CMS service to fetch the static assets from.

pomegranited committed:
Refactored to remove code duplication

Also added FIXME about nested vs separate child block OLX

symbolist commented on openedx_blockstore_relay/management/commands/transfer_to_blockstore.py:
```diff
@@ -0,0 +1,95 @@
+            help='UUID of the destination Blockstore Bundle, '
+                 'e.g., "01234567-89ab-cdef-fedc-ba9876543210"'
+        )
+        self.args['asset_root_url'] = parser.add_argument(
```
If we are running the command on an instance there are going to be production settings available, so in which case would be want this to be configurable?

pomegranited replied:
e675556 pulls the asset content directly from contentstore, so this argument is no longer needed.

symbolist commented on openedx_blockstore_relay/management/commands/transfer_to_blockstore.py:
```diff
@@ -0,0 +1,95 @@
+        """
+        Add named arguments.
+        """
+        self.args['block'] = parser.add_argument(
```
Nit: `block_key` or `block_usage_key` would be nicer. It is always confusing to me when a command parameter seems to expect a complex object rather than a simple type.

pomegranited replied:
Done, cf 83cc543

symbolist commented on openedx_blockstore_relay/management/commands/transfer_to_blockstore.py:
```diff
@@ -0,0 +1,95 @@
+        self.args['bundle'] = parser.add_argument(
+            '--bundle',
+            type=str,
+            required=True,
```
This would require creating a Bundle in the Blockstore each time separately. Would be simpler to pass a collection_uuid here and the command can create the Bundle itself.

pomegranited replied:
Or is it better to take a `collection_uuid`, and create a new Bundle for each child of the given `block`?

For example, if you specify a collection for each of these block types, you would get:
* block `type@course` --> Section (`chapter`) in each Bundle
* block `type@chapter` --> Subsection (`sequential`) in each Bundle
* block `type@sequential` --> Unit (`vertical`) in each Bundle.
* block `type@vertical` --> Subunit (`html`, `problem`, `video`) in each Bundle.

symbolist replied:
I would recommend keeping it simple for now and just putting everything in one Bundle. This will have to be refactored for the data transfer automation later anyway once there is more clarity about the product requirements.

pomegranited replied:
Ok cool, thanks @symbolist . I'll make the command accept *either* a `--bundle--uuid` (like it does currently) or a `--collection--uuid` (which will auto-create a target bundle), and import everything into that single bundle.

pomegranited replied:
Done with 0d302cd

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,257 @@
+        # TODO: handle full descendent tree to support these block types.
+        raise NotImplementedError("Block type not supported: %s" % block_type)
+    else:
+        transfer_obj = TransferBlock(block_key, bundle_uuid, asset_root_url)
```
Any reason why we can't use `TransferUnit` here too? For blocks without children the `for child in block.get_children():` loop should get skipped. Would be simpler to have generic code which descends the tree if there are child nodes and upload them one by one. It should automatically work for sequentials too. Something like:

```
def transfer_block_to_bundle():
  for child_block in block.get_children():
    transfer_child_block = TransferBlock(child_block)
    transfer_child_block.transfer_block_to_bundle()
  self.upload_assets()
  self.upload_olx()
```

pomegranited replied:
Yep, recursion is totally doable -- I was just holding off to limit the scope of this change 😄 

9e4e181 implements this, but it's not quite as clean as what you described, because:
* We don't want to update the `bundle.json` manifest file when transferring child blocks.
* We need to prevent the infinite recursion caused by Directed Acyclic Graphs.

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,257 @@
+        """
+        Upload any files found in the filesystem.
+        """
+        for item in filesystem.walk():
```
Can you explain what is happening here? Also what is being uploaded?

pomegranited replied:
I've renamed this method to make it clearer, and expanded on the comments.  Does that help?

cf https://github.com/open-craft/openedx-blockstore-relay/pull/1/commits/9e4e18121e377c293be33576637fe5f1f07fdf53#diff-7e4ed6a1b9b376cb82bb1a6d219be431R149

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,257 @@
+            self.upload_components(filesystem)
+            # FIXME -- either upload the child OLX, or append it to the root node.  Not both.
+            # self.upload_olx(child_olx, child)
+            olx_node.append(child_olx)
```
All the olx files should be separate in Blockstore.

pomegranited replied:
Thought of another reason why we have to maintain separate OLX files and just reference the children instead of inlining them -- Directed Acyclic Graphs!

pomegranited replied:
Fixed with recursion change: 9e4e181

bradenmacdonald replied:
@pomegranited 

> Thought of another reason why we have to maintain separate OLX files and just reference the children instead of inlining them -- Directed Acyclic Graphs!

Wait, which way do you mean that? For courses, we would like to enforce trees only and prohibit DAGs (by which I mean we don't want to allow DAGs where one block has multiple parents). They caused too many problems. 

https://openedx.atlassian.net/wiki/spaces/EDUCATOR/pages/34734086/Why+is+the+Course+Structure+a+DAG

https://openedx.atlassian.net/browse/TNL-4922

That means that if a course is a DAG, trying to import it into Blockstore as-is should fail with an error.

*Or* we could allow DAGs but if a block appears in multiple places, it gets treated by the LMS as two separate blocks. I think that's actually a better approach and supports adaptive learning and a bunch of other features. In that case, the fix would be in the XBlock runtime when it generates the full usage_key for each block, ensuring that the usage_key is based in part on the location of the block in the course.

pomegranited replied:
@bradenmacdonald 

Lol.. now I have no idea what this function should do when it encounters a DAG 😄 

Currently, the content that gets imported into the Bundle is as close to the Open edX Studio export as I could manage without creating a `runtime`.  The import loop checks to make sure that it's not recursing infinitely through DAGs, but it allows them without error, just like Studio does.  

We'd still need these checks in place to throw an error when a DAG is encountered, so if we decide later to disallow DAGs, we can do that.  Or if we decide that the DAGs should be allowed, but we want the sub-blocks inlined in a single OLX file that uses XIncludes or some other way to reference internal blocks, we can do that too.

Then how the LMS chooses to read all of this -- that's also up for discussion 😄

bradenmacdonald replied:
haha ok, thanks @pomegranited

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,257 @@
+
+        response = self._post_bundle_file(path=path, name=name, data=data, content_type=content_type)
+
+        if store_component:
```
This should happen in `upload_olx()` for the root `usage_key` that was passed to the command.

pomegranited replied:
Yep, agreed.  cf https://github.com/open-craft/openedx-blockstore-relay/pull/1/commits/9e4e18121e377c293be33576637fe5f1f07fdf53#diff-7e4ed6a1b9b376cb82bb1a6d219be431R168

symbolist reviewed (changes requested):
@pomegranited I have done a round of review. The main change that I think is needed is to make the transfer code walk recursively down the block tree. It should simplify the logic and remove most of the block specific code. Let me know if you think differently about it.

pomegranited committed:
Pulls the static asset content directly from contentstore

instead of requiring an asset-root-url and a running LMS/Studio.

Also moves the edx-platform imports into a `compat` module to facilitate testing.

pomegranited committed:
Recurse through the child blocks to allow for nested sequences.

* prevents infinite recursion (DAGs) by tracking which blocks have been processed.
* obviates the need for the separate TransferUnit class
* uses wrapped filesystem to prevent errors when processing video transcripts
* addresses PR nits

pomegranited committed:
Change management command arguments:

* from --block to --block-key
* from --bundle to --bundle-uuid

pomegranited committed:
Handled errors seen when uploading the DemoX course:

* AttributeError thrown for some XBlocks that don't provide definition_to_xml
* Empty file uploads cause issues, so upload a single space instead.

pomegranited commented:
Thank you for your review @symbolist !

I've addressed all your comments except for [this one](https://github.com/open-craft/openedx-blockstore-relay/pull/1#discussion_r213982682), where I have a question about what the desired behaviour should be.

pomegranited commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,233 @@
+                                        '/'.join(['bundles', str(self.bundle_uuid), 'files', '']))
+        self.manifest = {
+            'meta': {'version': self.BUNDLE_VERSION},
+            'type': self.olx_type(self.block_key.block_type),
```
cf https://github.com/open-craft/blockstore/pull/5#issuecomment-417362667
>  Use "olx/unit", "olx/collection", "olx/course", and perhaps "static"

pomegranited replied:
done, cf 4bd5b57

pomegranited commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,233 @@
+                    path=asset_path,
+                    data=content.data,
+                    content_type=content.content_type,
+                )
```
cf https://github.com/open-craft/blockstore/pull/5#discussion_r214021934:

Use `public=True` only for asset files, default for everything else should be `False`.

pomegranited replied:
Done, cf d5eb6e4 .
@symbolist Can you confirm that I understood this correctly?

symbolist replied:
Yup!

pomegranited commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,233 @@
+        self.bundle_files_url = urljoin(settings.BLOCKSTORE_API_URL,
+                                        '/'.join(['bundles', str(self.bundle_uuid), 'files', '']))
+        self.manifest = {
+            'meta': {'version': self.BUNDLE_VERSION},
```
cf https://github.com/open-craft/blockstore/pull/5#discussion_r214072438: change to `"schema": 0.1`

pomegranited replied:
Done cf 4a8d81b

bradenmacdonald commented:
😎

This is exciting to see! Thanks @pomegranited

symbolist commented on openedx_blockstore_relay/compat.py:
```diff
@@ -0,0 +1,75 @@
+
+    try:
+        asset_key = StaticContent.get_asset_key_from_path(course_key, asset_path)
+        return AssetManager.find(asset_key)
```
Oh, this looks so nice!

pomegranited replied:
Thanks 😄 Borrowed the idea for this module from Cliff's [`completion_aggregator.compat`](https://github.com/open-craft/openedx-completion-aggregator/blob/master/completion_aggregator/compat.py).

symbolist commented on openedx_blockstore_relay/management/commands/transfer_to_blockstore.py:
```diff
@@ -0,0 +1,81 @@
+        Add named arguments.
+        """
+        self.args['block_key'] = parser.add_argument(
+            '--block-key',
```
💯

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,233 @@
+        """
+        Upload any files stored in the filesystem during OLX generation.
+
+        Mostly (exclusively?) consists of `.html` files associated with HTML blocks.
```
This will have to be per block type ultimately.

pomegranited replied:
I've updated the comment to reflect this note, cf https://github.com/open-craft/openedx-blockstore-relay/commit/8f781ced07cad132d8989537d210c66fe981b9f4#diff-7e4ed6a1b9b376cb82bb1a6d219be431R156

symbolist commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,233 @@
+
+        block_type = block_key.block_type
+        if block_type in ('course', 'chapter', 'sequential', 'vertical'):
+            # These blocks' definition_to_xml methods require a runtime, so we avoid calling them.
```
Fine for now, but ultimately we will have to do that.

pomegranited replied:
Noted that in the comment: cf https://github.com/open-craft/openedx-blockstore-relay/commit/8f781ced07cad132d8989537d210c66fe981b9f4#diff-7e4ed6a1b9b376cb82bb1a6d219be431R95

symbolist reviewed (changes requested):
Unfortunately, this is not working for the video transcripts because in the xml the transcript filename is without `/static/` prepended to it. For example:

```
<video youtube="1.00:5UbwgHKP5qo" url_name="8c81b025c85d4fa19a6a4d1ff4bf19bd" display_name="" download_video="false" edx_video_id="bb5e20cf-654e-4dc0-a6ac-8fa7b9291d3d" html5_sources="[]" youtube_id_1_0="5UbwgHKP5qo">
  <transcript language="en" src="bb5e20cf-654e-4dc0-a6ac-8fa7b9291d3d-en.srt"/>
</video>
```

So these will have to be read from the xml or block `transcripts` field and then files explicitly fetched from the content store. `_upload_related_files()` may be the right place to do it for now. Eventually, the per-block processing will have to be moved to a method on the block themselves.

Otherwise everything else looks close to done. The generic tree traversal has made the code clearer too!

pomegranited committed:
Fixed the manifest bundle.json `type` field

according to https://github.com/open-craft/blockstore/blob/symbolist/bundle-format/docs/bundles.rst#type

pomegranited committed:
Use `public=False` for bundle files, except assets.

pomegranited committed:
Store `schema` in bundle.json, not `meta.version`

pomegranited committed:
Updates comments to reflect Usman's notes on PR

pomegranited committed:
Separates the "bundle type" from the "content type"

and makes these private methods, so I don't have to test them :)

pomegranited commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -0,0 +1,233 @@
+        Upload the collected static asset files to the bundle.
+        """
+        for (asset_path, asset_uri) in self.assets:
+            content = compat.get_asset_content_from_path(self.block_key.course_key, asset_path)
```
@symbolist Do we want [asset `thumbnails`](https://github.com/edx/edx-platform/blob/b85d6f031c1d6590a184a873fcadb3b2d42024bd/common/lib/xmodule/xmodule/contentstore/content.py#L71-L86) too?

symbolist replied:
Hmm. We do not have any place to store this metadata yet so this can be added later.

pomegranited committed:
Adds support for pulling video transcripts out of video blocks

Also refactors the way we handle static assets -- since the existing video
xmodule code reads the transcript from the contentstore, then it made sense to
start uploading asset files as they're read instead of storing them in memory
to be uploaded later.

pomegranited committed:
Accepts a --collection-uuid

Creates a new bundle on that collection, and uploads the block content there.

pomegranited commented:
@symbolist  I think I've addressed everything from your review now, please see replies inline and the updated PR description.

5245ba8 should now pull all available transcript files for a video and store them to the bundle. I still can't test this though -- need to upgrade my devstack.  Do you mind to test again?

No tests added yet either.. hope to have at least a minimal set added next week, unless the upcoming bundle format changes make this unnecessary?

symbolist committed:
Add tests.

symbolist commented:
@pomegranited Thanks! I have added the tests. I also checked that the transcripts are being extracted and transferred correctly. Though for the production course exports that we are also using for testing, for video blocks whose assets are stored using edxval on the edX production servers, are skipped when running the command in the devstack.

- [x] I tested this by transferring a number of units from the test courses to the Blockstore.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~
- [x] Includes documentation

```
=================================================================================================== test session starts ====================================================================================================
platform linux2 -- Python 2.7.12, pytest-3.6.3, py-1.6.0, pluggy-0.6.0
Using --randomly-seed=1535830371
Django settings: test_settings (from ini file)
rootdir: /edx/src/openedx-blockstore-relay, inifile: tox.ini
plugins: xdist-1.23.0, randomly-1.2.3, forked-0.2, django-3.1.2, cov-2.5.1, attrib-0.1.3
collected 3 items                                                                                                                                                                                                          

openedx_blockstore_relay/management/commands/tests/test_transfer_to_blockstore.py .                                                                                                                                  [ 33%]
openedx_blockstore_relay/tests/test_transfer_data.py ..                                                                                                                                                              [100%]

---------- coverage: platform linux2, python 2.7.12-final-0 ----------
Name                                                                                Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------------------------------------------------------------------------------
openedx_blockstore_relay/__init__.py                                                    3      0      0      0   100%
openedx_blockstore_relay/apps.py                                                        8      0      0      0   100%
openedx_blockstore_relay/compat.py                                                     63     54     12      0    12%   32-33, 40-45, 54-66, 73-85, 96-128
openedx_blockstore_relay/management/__init__.py                                         0      0      0      0   100%
openedx_blockstore_relay/management/commands/__init__.py                                0      0      0      0   100%
openedx_blockstore_relay/management/commands/tests/test_transfer_to_blockstore.py      29      0      0      0   100%
openedx_blockstore_relay/management/commands/transfer_to_blockstore.py                 52      5     12      1    84%   100-104, 98->100
openedx_blockstore_relay/models.py                                                      2      0      0      0   100%
openedx_blockstore_relay/settings/__init__.py                                           0      0      0      0   100%
openedx_blockstore_relay/settings/common.py                                            16     16      2      0     0%   5-37
openedx_blockstore_relay/settings/production.py                                         2      2      0      0     0%   5-7
openedx_blockstore_relay/test_utils/__init__.py                                         0      0      0      0   100%
openedx_blockstore_relay/test_utils/compat.py                                          20      2      4      2    83%   21, 26, 19->21, 24->26
openedx_blockstore_relay/tests/test_transfer_data.py                                   56      0      4      0   100%
openedx_blockstore_relay/transfer_data.py                                             126     10     34      7    89%   113, 128-131, 164, 166, 199-200, 240, 35->37, 111->113, 124->128, 163->164, 165->166, 198->199, 238->240
openedx_blockstore_relay/urls.py                                                        5      5      0      0     0%   2-10
-------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                 382     94     68     10    73%
```

bradenmacdonald commented on openedx_blockstore_relay/tests/test_transfer_data.py:
```diff
@@ -0,0 +1,128 @@
+from openedx_blockstore_relay.test_utils.compat import StubCompat, StaticContent
+
+
+class TestXBlock(XBlock):
```
Nit: a docstring would be helpful here. At first glance it's not clear if this is "an XBlock for use in tests" or "a tool to help test an XBlock"

symbolist replied:
Added.

bradenmacdonald commented on openedx_blockstore_relay/tests/test_transfer_data.py:
```diff
@@ -0,0 +1,128 @@
+
+        with mock.patch('openedx_blockstore_relay.transfer_data.requests') as mock_requests:
+            with mock.patch('openedx_blockstore_relay.transfer_data.TransferBlock.transfer_block_to_bundle'):
+                transfer_to_blockstore(self.block_key, collection_uuid=self.COLLECTION_UUID)
```
Can we add any asserts here or do we just need to test that no exceptions occurred?

symbolist replied:
Right now if an error happens during uploads of the assets, there is no rollback. We will have to wait for the draft/publish workflow to be able to add that. So cannot see how having this function return something will be useful. For now as long as there are no exceptions this succeeded.

bradenmacdonald commented on openedx_blockstore_relay/tests/test_transfer_data.py:
```diff
@@ -0,0 +1,128 @@
+            mock_response.json.return_value = {
+                'path': '/vertical/v1.olx',
+            }
+            mock_requests.post.return_value = mock_response
```
Can you add a comment saying which POST request this is mocking? from the code here I can't actually tell what API endpoint is being mocked out, just that it's some POST request.

bradenmacdonald reviewed (approved):
👍 @symbolist The tests look great! I left some very minor comments for your consideration but I don't need to review them again.

- [ ] ~~I tested this~~: n/a, relying on your testing. This was just a quick review of the last commit which is only test code.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ n/a
- [ ] ~~Includes documentation~~ n/a

pomegranited commented:
@symbolist Thank you for adding these tests!

> Though for the production course exports that we are also using for testing, for video blocks whose assets are stored using edxval on the edX production servers, are skipped when running the command in the devstack.

Ok, I was under the impression that only contentstore assets needed to be copied over, and so have been focusing on these, not externally-hosted resources like CDN scripts/images or S3 hosted videos.  Should we be pulling those in as well?  I can create a follow-up task to address this -- we'll need to change the external references in OLX to point to the new Bundle locations.

symbolist committed:
Added comments.

symbolist commented:
@pomegranited I think there is some discovery work that will be needed for assets. Studio allows for "locking" assets and then there can be derived assets like thumbnails and video formats. So all this will have be thought about holistically before support can added to the transfer mechanism.

pomegranited commented:
👍 Thanks @symbolist 😄
